py4DSTEM/process/phase/utils.py

```diff
@@ -1674,3 +1675,547 @@ def rotate_point(origin, point, angle):
     qx = ox + np.cos(angle) * (px - ox) - np.sin(angle) * (py - oy)
     qy = oy + np.sin(angle) * (px - ox) + np.cos(angle) * (py - oy)
     return qx, qy
+
+
+def bilinearly_interpolate_array(
+    image,
+    xa,
+    ya,
+    xp=np,
+):
+    """
+    Bilinear sampling of intensities from an image array and pixel positions.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        Image array to sample from
+    xa: np.ndarray
+        Vertical positions of image array in pixels
+    ya: np.ndarray
+        Horizontal positions of image array in pixels
+
+    Returns
+    -------
+    intensities: np.ndarray
+        Bilinearly-sampled intensities of array at (xa,ya) positions
+
+    """
+
+    xF = xp.floor(xa).astype("int")
+    yF = xp.floor(ya).astype("int")
+    dx = xa - xF
+    dy = ya - yF
+
+    all_inds = [
+        [xF, yF],
+        [xF + 1, yF],
+        [xF, yF + 1],
+        [xF + 1, yF + 1],
+    ]
+
+    all_weights = [
+        (1 - dx) * (1 - dy),
+        (dx) * (1 - dy),
+        (1 - dx) * (dy),
+        (dx) * (dy),
+    ]
+
+    raveled_image = image.ravel()
+    intensities = xp.zeros(xa.shape, dtype=xp.float32)
+    # filter_weights = xp.zeros(xa.shape, dtype=xp.float32)
+
+    for inds, weights in zip(all_inds, all_weights):
+        intensities += (
+            raveled_image[
+                xp.ravel_multi_index(
+                    inds,
+                    image.shape,
+                    mode=["wrap", "wrap"],
+                )
+            ]
+            * weights
+        )
+        # filter_weights += weights
+
+    return intensities  # / filter_weights # unnecessary, sums up to unity
+
+
+def lanczos_interpolate_array(
+    image,
+    xa,
+    ya,
+    alpha,
+    xp=np,
+):
+    """
+    Lanczos sampling of intensities from an image array and pixel positions.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        Image array to sample from
+    xa: np.ndarray
+        Vertical positions of image array in pixels
+    ya: np.ndarray
+        Horizontal positions of image array in pixels
+    alpha: int
+        Lanczos kernel order
+
+    Returns
+    -------
+    intensities: np.ndarray
+        Lanczos-sampled intensities of array at (xa,ya) positions
+
+    """
+    xF = xp.floor(xa).astype("int")
+    yF = xp.floor(ya).astype("int")
+    dx = xa - xF
+    dy = ya - yF
+
+    all_inds = []
+    all_weights = []
+
+    for i in range(-alpha + 1, alpha + 1):
+        for j in range(-alpha + 1, alpha + 1):
```

Comment:
Is this sum meant to be asymmetric about zero? i.e. should it be -alpha to alpha inclusive, or -alpha+1 to alpha as written?

Reply:
This is correct as-written, the interpolation weights are computed on a [0,1] grid (same as we do for bilinear). See equation here: https://pixinsight.com/doc/docs/InterpolationAlgorithms/InterpolationAlgorithms.html